File: integration-test/src/test/js/cnft-util.js
const { webcrypto } = require('crypto');
const { ec: EC } = require('elliptic');

const {
    CNFT,
    CNFTCrypto,
    DNA,
    DescriptionField,
    FieldMeta,
    FieldType,
    Gene,
    GeneID,
    HDPathStore,
    TokenTypeMeta,
} = require('wallet-lib');

const Ascii = {
    Reset: '\x1b[0m',

    FgGreen: '\x1b[32m',
    FgYellow: '\x1b[33m',
    FgMagenta: '\x1b[35m',
};

const CNFTGateUrl = 'http://localhost:8981';
const CNFTRemoteWalletUrl = 'http://localhost:8983';
const Chain = CNFT.connect(CNFTGateUrl);

function log(msg) {
    console.log(msg);
}

function toB64(arrayBuffer) {
    return btoa(String.fromCharCode(...new Uint8Array(arrayBuffer)));
}

async function createWalletCrypto() {
    const jsec = new EC('secp256k1');
    return await CNFTCrypto
        .newContext(
            /* identityOpsFactory = */ CNFTCrypto.hdSignatureOperations(webcrypto, jsec, HDPathStore.inMemory()),
            /* encryptionOpsFactory = */ CNFTCrypto.webCryptoEncryption(webcrypto),
            /* accessOpsFactory = */ CNFTCrypto.webCryptoAccessOperations(webcrypto, jsec),
            /* addressOpsFactory = */ CNFTCrypto.hdSignatureOperations(webcrypto, jsec, HDPathStore.inMemory()),
            /* hashFactory = */ CNFTCrypto.hash(webcrypto),
            CNFTCrypto.secureRandomGenerator(webcrypto)
        )
        .create();
}

async function registerMemberHook(crypto) {
    const { FgGreen: FgBase, FgMagenta: FgVal, Reset } = Ascii;
    log(`${FgBase}Connecting to Admin wallet...${Reset}`);

    const adminWallet = await Chain.connectWallet(CNFTRemoteWalletUrl);

    const memberInfo = await crypto.memberInformation();

    try {
        await Chain.getMember(memberInfo.id);
        log(`${FgBase}Member already registered${Reset}`);
    } catch (error) {
        log(`${FgBase}Member is not registered yet (${error})${Reset}`);
        const tx = await adminWallet.registerMember(memberInfo);
        log(`${FgBase}Registered self [${FgVal}${memberInfo.id}${FgBase}] in TX: [${FgVal}${tx.blockNumber}` +
            `${FgBase} : ${FgVal}${tx.txId}${FgBase}]${Reset}`);
    }
}

//

async function registerTokenType(wallet, tokenTypeName, dnaChange = []) {
    const typeId = await wallet.createId;

    await wallet.registerTokenType(
        typeId,
        TokenTypeMeta(
            [ DescriptionField('Value', FieldType.Text, tokenTypeName) ],
            [ FieldMeta('amount', FieldType.Text) ]
        ),
        DNA(
            [ Gene(GeneID.EmissionControlledByIssuer, []) ],
            [ Gene(GeneID.RequireOwnerSignatureForTransfer, []) ],
            [ Gene(GeneID.RequireOwnerSignatureForBurn, []) ],
            dnaChange
        ),
        [],
        []
    );

    return typeId;
}

module.exports = {
    Ascii,
    createWalletCrypto,
    Chain,
    log,
    registerMemberHook,
    registerTokenType,
    toB64,
};

File: integration-test/src/test/scala/ru/sberbank/blockchain/cnft/wallet/test/CNFTUtil.scala
package ru.sberbank.blockchain.cnft.wallet.test

import ru.sberbank.blockchain.cnft.common.types.{Collection, Result}
import ru.sberbank.blockchain.cnft.commons.LoggingSupport
import ru.sberbank.blockchain.cnft.gate.model.TxResult
import ru.sberbank.blockchain.cnft.model.{DNA, DescriptionField, FieldMeta, FieldType, Gene, GeneID, TokenTypeMeta}
import ru.sberbank.blockchain.cnft.wallet.crypto.WalletCrypto
import ru.sberbank.blockchain.cnft.wallet.dsl._
import ru.sberbank.blockchain.cnft.wallet.spec.{CNFTFactory, CNFTWalletSpec}
import ru.sberbank.blockchain.cnft.wallet.store.HDPathStore
import ru.sberbank.blockchain.cnft.wallet.{CNFT, CNFTCrypto}

object CNFTUtil extends LoggingSupport {

    lazy val CNFTGateUrl: String = sys.env.getOrElse("CNFT_GATE_URL", "http://localhost:8981")

    val CNFTRemoteWalletUrl: String = "http://localhost:8983"

    val Chain: CNFTFactory[Result] = CNFT.connect(CNFTGateUrl)


    def createWalletCrypto: WalletCrypto[Result] =
        CNFTCrypto
            .newContext(
                identityOpsFactory = CNFTCrypto.hdSignatureOperations(HDPathStore.inMemory()),
                encryptionOpsFactory = CNFTCrypto.bouncyCastleEncryption(),
                accessOpsFactory = CNFTCrypto.bouncyCastleAccessOperations(),
                addressOpsFactory = CNFTCrypto.hdSignatureOperations(HDPathStore.inMemory()),
                hashFactory = CNFTCrypto.hash(),
                randomGeneratorFactory = CNFTCrypto.secureRandomGenerator()
            )
            .create()
            .orFail("Unable to create WalletCrypto")

    def registerMemberHook(crypto: WalletCrypto[Result]): Unit = {
        logger.info("Connecting to Admin wallet...")

        val adminWallet = Chain.connectWallet(CNFTRemoteWalletUrl)

        crypto.memberInformation().map { memberInfo =>
            Chain.getMember(memberInfo.id) match {
                case Left(_) =>
                    logger.info("Member does not registered yet")
                    val tx = adminWallet
                        .registerMember(memberInfo)
                        .orFail("Failed to register self in blockchain")
                    logger.info(s"Registered self [${memberInfo.id}] in TX: [${tx.blockNumber} : ${tx.txId}]")

                case Right(_) =>
                    logger.info("Member already registered")
            }
        }.orFail("Failed to register member")
    }

    def registerEmittableTokenType(
        wallet: CNFTWalletSpec[Result],
        typeId: String,
        tokenTypeName: String,
        dnaChange: Collection[Gene]
    ): TxResult[Unit] =
        wallet
            .registerTokenType(
                typeId,
                TokenTypeMeta(
                    description = Collection(DescriptionField("Value", FieldType.Text, tokenTypeName)),
                    fields = Collection(FieldMeta("amount", FieldType.Text))
                ),
                DNA.defaultInstance.copy(
                    emission = Collection(
                        Gene(
                            id = GeneID.EmissionControlledByIssuer,
                            parameters = Collection.empty
                        )
                    ),
                    change = dnaChange
                ),
                Collection.empty,
                Collection.empty
            )
            .orFail(s"Failed to register fungible token type $tokenTypeName")

}
